File: src/lua/meth/mat2.lua
---@class meth.Mat2
---@field x1 float
---@field y1 float
---@field x2 float
---@field y2 float
---@operator add(meth.Mat2): meth.Mat2
---@operator add(number): meth.Mat2
---@operator sub(meth.Mat2): meth.Mat2
---@operator sub(number): meth.Mat2
---@operator mul(meth.Mat2): meth.Mat2
---@operator mul(meth.Vec2): meth.Vec2
---@operator mul(number): meth.Mat2

---@class meth.Mat2
local methods = {
	---@param self meth.Mat2
	---@param value number|meth.Mat2
	---@return meth.Mat2
	add = function(self, value)
		if type(value) == "number" then
			self[1] = self[1] + value
			self[2] = self[2] + value
			self[3] = self[3] + value
			self[4] = self[4] + value
		else
			self[1] = self[1] + value[1]
			self[2] = self[2] + value[2]
			self[3] = self[3] + value[3]
			self[4] = self[4] + value[4]
		end
		return self
	end,
	---@param self meth.Mat2
	---@param value number|meth.Mat2
	---@return meth.Mat2
	sub = function(self, value)
		if type(value) == "number" then
			self[1] = self[1] - value
			self[2] = self[2] - value
			self[3] = self[3] - value
			self[4] = self[4] - value
		else
			self[1] = self[1] - value[1]
			self[2] = self[2] - value[2]
			self[3] = self[3] - value[3]
			self[4] = self[4] - value[4]
		end
		return self
	end,
	---@param self meth.Mat2
	---@param value number|meth.Mat2
	---@return meth.Mat2
	mul = function(self, value)
		if type(value) == "number" then
			self[1] = self[1] * value
			self[2] = self[2] * value
			self[3] = self[3] * value
			self[4] = self[4] * value
		else
			---@cast value meth.Mat2
			local a, c, b, d = self[1], self[2], self[3], self[4]
			local e, g, f, h = value[1], value[2], value[3], value[4]
			self[1] = a * e + b * g
			self[2] = c * e + d * g
			self[3] = a * f + b * h
			self[4] = c * f + d * h
		end
		return self
	end,
	type = "Mat2",
}

local metatable = {
	__newindex = function(self, key, value)
		if key == "x1" then
			self[1] = value
		elseif key == "y1" then
			self[2] = value
		elseif key == "x2" then
			self[3] = value
		elseif key == "y2" then
			self[4] = value
		end
	end,
	---@param self meth.Vec2
	__index = function(self, value)
		if value == "x1" then
			return self[1]
		elseif value == "y1" then
			return self[2]
		elseif value == "x2" then
			return self[3]
		elseif value == "y2" then
			return self[4]
		else
			return methods[value]
		end
	end,
	__add = function(a, b)
		if type(a) == "number" then
			return mat2(a + b[1], a + b[2], a + b[3], a + b[4])
		elseif type(b) == "number" then
			return mat2(a[1] + b, a[2] + b, a[3] + b, a[4] + b)
		else
			return mat2(a[1] + b[1], a[2] + b[2], a[3] + b[3], a[4] + b[4])
		end
	end,
	__sub = function(a, b)
		if type(a) == "number" then
			return mat2(a - b[1], a - b[2], a - b[3], a - b[4])
		elseif type(b) == "number" then
			return mat2(a[1] - b, a[2] - b, a[3] - b, a[4] - b)
		else
			return mat2(a[1] - b[1], a[2] - b[2], a[3] - b[3], a[4] - b[4])
		end
	end,
	__mul = function(a, b)
		if type(a) == "number" then
			return mat2(a * b[1], a * b[2], a * b[3], a * b[4])
		elseif type(b) == "number" then
			return mat2(a[1] * b, a[2] * b, a[3] * b, a[4] * b)
		elseif type(b) == "table" and b.type == "Vec2" then
			local x, y = b[1], b[2]
			local a, b, c, d = a[1], a[2], a[3], a[4]
			return vec2(a * x + c * y, b * x + d * y)
		else
			local a, c, b, d = a[1], a[2], a[3], a[4]
			local e, g, f, h = b[1], b[2], b[3], b[4]
			return mat2(a * e + b * g, c * e + d * g, a * f + b * h, c * f + d * h)
		end
	end,
	__tostring = function(self)
		return "(x_axis: (" .. self[1] .. ", " .. self[2] .. "); y_axis: (" .. self[3] .. ", " .. self[4] .. "))"
	end,
}

---@return meth.Mat2
---@overload fun(): meth.Mat2
---@overload fun(value: number): meth.Mat2
---@overload fun(value: meth.Vec4): meth.Mat2
---@overload fun(x_axis: meth.Vec2, y_axis: meth.Vec2): meth.Mat2
---@overload fun(x1: number, y1: number, x2: number, y2: number): meth.Mat2
function mat2(x1, y1, x2, y2)
	if type(x1) == "table" and x1.type and type(y1) == "table" and y1.type then
		local x1_type, y1_type = x1.type, y1.type
		if (x1_type == "Vec2" or x1_type == "IVec2") and (y1_type == "Vec2" or y1_type == "IVec2") then
			---@cast x1 meth.Vec2
			---@cast y1 meth.Vec2
			return setmetatable({ x1.x, x1.y, y1.x, y1.y }, metatable)
		elseif x1_type == "Vec4" or x1_type == "IVec4" then
			---@cast x1 meth.Vec4
			return setmetatable({ x1.x, x1.y, x1.z, x1.w }, metatable)
		end
		return setmetatable({ 0, 0, 0, 0 }, metatable)
	elseif x1 and y1 and x2 and y2 then
		return setmetatable({ x1, y1, x2, y2 }, metatable)
	elseif x1 and not y1 then
		return setmetatable({ x1, x1, x1, x1 }, metatable)
	else
		return setmetatable({ 0, 0, 0, 0 }, metatable)
	end
end

local inner = _G["__inner"]
---@type __inner
local inner = inner ~= nil and inner or {}
_G["__inner"] = inner
inner.mat2_metatable = metatable

---@class meth.Mat2.definitions
local Mat2 = {
	---@return meth.Mat2
	---@param angle number (in radians)
	from_angle = function(angle)
		local sin, cos = math.sin(angle), math.cos(angle)
		return mat2(cos, -sin, sin, cos)
	end,
	ZERO = mat2(),
	IDENTITY = mat2(1, 0, 0, 1),
}

return Mat2

File: src/mat2.rs
use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};

use crate::vec2::Vec2;

#[derive(Default, Debug, PartialEq, Clone, Copy, serde::Serialize, serde::Deserialize)]
pub struct Mat2 {
    pub x_axis: Vec2,
    pub y_axis: Vec2,
}

impl Mat2 {
    pub const ZERO: Self = Self::splat(0.0);
    pub const IDENTITY: Self = Self::new(1.0, 0.0, 0.0, 1.0);
    pub const fn new(x0: f32, y0: f32, x1: f32, y1: f32) -> Self {
        Self {
            x_axis: Vec2::new(x0, y0),
            y_axis: Vec2::new(x1, y1),
        }
    }
    pub const fn from_axis(x_axis: Vec2, y_axis: Vec2) -> Self {
        Self { x_axis, y_axis }
    }
    pub const fn splat(value: f32) -> Self {
        Self::from_axis(Vec2::splat(value), Vec2::splat(value))
    }
    pub fn from_angle(angle: f32) -> Self {
        let (sin, cos) = f32::sin_cos(angle);
        Self::new(cos, -sin, sin, cos)
    }
}

impl Add for Mat2 {
    type Output = Mat2;

    fn add(self, rhs: Self) -> Self::Output {
        Self {
            x_axis: self.x_axis + rhs.x_axis,
            y_axis: self.y_axis + rhs.y_axis,
        }
    }
}

impl AddAssign<Mat2> for Mat2 {
    fn add_assign(&mut self, rhs: Mat2) {
        self.x_axis += rhs.x_axis;
        self.y_axis += rhs.y_axis;
    }
}

impl Add<f32> for Mat2 {
    type Output = Mat2;

    fn add(self, rhs: f32) -> Self::Output {
        Self {
            x_axis: self.x_axis + rhs,
            y_axis: self.y_axis + rhs,
        }
    }
}

impl AddAssign<f32> for Mat2 {
    fn add_assign(&mut self, rhs: f32) {
        self.x_axis += rhs;
        self.y_axis += rhs;
    }
}

impl Sub for Mat2 {
    type Output = Mat2;

    fn sub(self, rhs: Self) -> Self::Output {
        Self {
            x_axis: self.x_axis - rhs.x_axis,
            y_axis: self.y_axis - rhs.y_axis,
        }
    }
}

impl SubAssign<Mat2> for Mat2 {
    fn sub_assign(&mut self, rhs: Mat2) {
        self.x_axis -= rhs.x_axis;
        self.y_axis -= rhs.y_axis;
    }
}

impl Sub<f32> for Mat2 {
    type Output = Mat2;

    fn sub(self, rhs: f32) -> Self::Output {
        Self {
            x_axis: self.x_axis - rhs,
            y_axis: self.y_axis - rhs,
        }
    }
}

impl SubAssign<f32> for Mat2 {
    fn sub_assign(&mut self, rhs: f32) {
        self.x_axis -= rhs;
        self.y_axis -= rhs;
    }
}

impl Mul<f32> for Mat2 {
    type Output = Mat2;

    fn mul(self, rhs: f32) -> Self::Output {
        Self {
            x_axis: self.x_axis * rhs,
            y_axis: self.y_axis * rhs,
        }
    }
}

impl MulAssign<f32> for Mat2 {
    fn mul_assign(&mut self, rhs: f32) {
        self.x_axis *= rhs;
        self.y_axis *= rhs;
    }
}

impl Mul for Mat2 {
    type Output = Mat2;

    fn mul(self, rhs: Self) -> Self::Output {
        let (a, c, b, d) = (self.x_axis.x, self.x_axis.y, self.y_axis.x, self.y_axis.y);
        let (e, g, f, h) = (rhs.x_axis.x, rhs.x_axis.y, rhs.y_axis.x, rhs.y_axis.y);
        Self {
            x_axis: Vec2 {
                x: a * e + b * g,
                y: c * e + d * g,
            },
            y_axis: Vec2 {
                x: a * f + b * h,
                y: c * f + d * h,
            },
        }
    }
}

impl MulAssign for Mat2 {
    fn mul_assign(&mut self, rhs: Self) {
        let (a, c, b, d) = (self.x_axis.x, self.x_axis.y, self.y_axis.x, self.y_axis.y);
        let (e, g, f, h) = (rhs.x_axis.x, rhs.x_axis.y, rhs.y_axis.x, rhs.y_axis.y);
        self.x_axis.x = a * e + b * g;
        self.x_axis.y = c * e + d * g;
        self.y_axis.x = a * f + b * h;
        self.y_axis.y = c * f + d * h;
    }
}

impl Mul<Vec2> for Mat2 {
    type Output = Vec2;

    fn mul(self, rhs: Vec2) -> Self::Output {
        let (x, y) = (rhs.x, rhs.y);
        let (a, b, c, d) = (self.x_axis.x, self.x_axis.y, self.y_axis.x, self.y_axis.y);
        Vec2 {
            x: a * x + c * y,
            y: b * x + d * y,
        }
    }
}

impl Div<f32> for Mat2 {
    type Output = Mat2;

    fn div(self, rhs: f32) -> Self::Output {
        Self {
            x_axis: self.x_axis / rhs,
            y_axis: self.y_axis / rhs,
        }
    }
}

impl DivAssign<f32> for Mat2 {
    fn div_assign(&mut self, rhs: f32) {
        self.x_axis /= rhs;
        self.y_axis /= rhs;
    }
}

impl mlua::FromLua for Mat2 {
    fn from_lua(value: mlua::Value, _lua: &mlua::Lua) -> mlua::Result<Self> {
        let table = value.as_table().ok_or_else(|| {
            lua_error::lua_error!(
                "`{}` should be a table, got `{}` instead",
                stringify!(Mat2),
                value.type_name()
            )
        })?;
        Ok(Mat2 {
            x_axis: Vec2 {
                x: mlua::ErrorContext::with_context(table.raw_get(1u32), |_| {
                    "could not get `Mat2`s field `x_axis.x` of type `f32`"
                })?,
                y: mlua::ErrorContext::with_context(table.raw_get(2u32), |_| {
                    "could not get `Mat2`s field `x_axis.y` of type `f32`"
                })?,
            },
            y_axis: Vec2 {
                x: mlua::ErrorContext::with_context(table.raw_get(3u32), |_| {
                    "could not get `Mat2`s field `y_axis.x` of type `f32`"
                })?,
                y: mlua::ErrorContext::with_context(table.raw_get(4u32), |_| {
                    "could not get `Mat2`s field `y_axis.y` of type `f32`"
                })?,
            },
        })
    }
}

impl mlua::IntoLua for Mat2 {
    fn into_lua(self, lua: &mlua::Lua) -> mlua::Result<mlua::Value> {
        let table = lua.create_table()?;
        let inner_table = lua
            .globals()
            .get::<Option<mlua::Table>>("__inner")?
            .ok_or_else(|| lua_error::lua_error!("could not get __inner"))?;
        let metatable = inner_table
            .get::<Option<mlua::Table>>("mat2_metatable")?
            .ok_or_else(|| lua_error::lua_error!("could not get metatable {}", "mat2_metatable"))?;
        table.set_metatable(Some(metatable));
        mlua::ErrorContext::with_context(table.raw_set(1i32, self.x_axis.x), |_| {
            "could not set `Mat2`s field `x_axis.x` of type `u32`"
        })?;
        mlua::ErrorContext::with_context(table.raw_set(2i32, self.x_axis.y), |_| {
            "could not set `Mat2`s field `x_axis.y` of type `u32`"
        })?;
        mlua::ErrorContext::with_context(table.raw_set(3i32, self.y_axis.x), |_| {
            "could not set `Mat2`s field `y_axis.x` of type `u32`"
        })?;
        mlua::ErrorContext::with_context(table.raw_set(4i32, self.y_axis.y), |_| {
            "could not set `Mat2`s field `y_axis.y` of type `u32`"
        })?;
        Ok(mlua::Value::Table(table))
    }
}
